cpp/src/arrow/filesystem/s3fs_benchmark.cc

```diff
@@ -0,0 +1,344 @@
+                              Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
+                              use_virtual_addressing));
+
+    MakeFileSystem();
+
+    const char* skip_str = std::getenv(kEnvSkipSetup);
+    const std::string skip = skip_str ? std::string(skip_str) : "";
+    if (skip == "true") {
+      std::cerr << "Skipping creation of bucket/objects as requested" << std::endl;
+    } else {
+      ASSERT_OK(MakeBucket());
+      ASSERT_OK(MakeObject("bytes_1mib", 1024 * 1024));
+      ASSERT_OK(MakeObject("bytes_100mib", 100 * 1024 * 1024));
+      ASSERT_OK(MakeObject("bytes_500mib", 500 * 1024 * 1024));
+      ASSERT_OK(MakeParquetObject(bucket_ + "/pq_c100_r250k", 100, 250000));
+    }
+  }
+
+  void MakeFileSystem() {
+    options_.ConfigureAccessKey(minio_.access_key(), minio_.secret_key());
+    options_.scheme = "http";
+    if (!region_.empty()) {
+      options_.region = region_;
+    }
+    options_.endpoint_override = minio_.connect_string();
+    ASSERT_OK_AND_ASSIGN(fs_, S3FileSystem::Make(options_));
+  }
+
+  /// Set up bucket if it doesn't exist.
+  ///
+  /// When using Minio we'll have a fresh setup each time, but
+  /// otherwise we may have a leftover bucket.
+  Status MakeBucket() {
+    Aws::S3::Model::HeadBucketRequest head;
+    head.SetBucket(ToAwsString(bucket_));
+    const Status st = OutcomeToStatus(client_->HeadBucket(head));
+    if (st.ok()) {
+      // Bucket exists already
+      return st;
+    }
+    Aws::S3::Model::CreateBucketRequest req;
+    req.SetBucket(ToAwsString(bucket_));
+    return OutcomeToStatus(client_->CreateBucket(req));
+  }
+
+  /// Make an object with dummy data.
+  Status MakeObject(const std::string& name, int size) {
+    Aws::S3::Model::PutObjectRequest req;
+    req.SetBucket(ToAwsString(bucket_));
+    req.SetKey(ToAwsString(name));
+    req.SetBody(std::make_shared<std::stringstream>(std::string(size, 'a')));
+    return OutcomeToStatus(client_->PutObject(req));
+  }
+
+  /// Make an object with Parquet data.
+  Status MakeParquetObject(const std::string& path, int num_columns, int num_rows) {
+    std::vector<std::shared_ptr<ChunkedArray>> columns(num_columns);
+    std::vector<std::shared_ptr<Field>> fields(num_columns);
+    for (int i = 0; i < num_columns; ++i) {
+      arrow::random::RandomArrayGenerator generator(i);
+      std::shared_ptr<Array> values = generator.Float64(num_rows, -1.e10, 1e10, 0);
+      std::stringstream ss;
+      ss << "col" << i;
+      columns[i] = std::make_shared<ChunkedArray>(values);
+      fields[i] = ::arrow::field(ss.str(), values->type());
+    }
+    auto schema = std::make_shared<::arrow::Schema>(fields);
+
+    std::shared_ptr<Table> table = Table::Make(schema, columns);
+
+    std::shared_ptr<io::OutputStream> sink;
+    ARROW_ASSIGN_OR_RAISE(sink, fs_->OpenOutputStream(path));
+    auto writer_properties =
+        parquet::WriterProperties::Builder().write_batch_size(128)->build();
```

Comment:
Does `write_batch_size` influence the row group size or something? 128 rows at once sounds tiny.

Reply:
I copied some of this out of the unit tests. I've changed this to just use defaults for everything.